### ISSUE TITLE
a question about the attention mask

### ISSUE THREAD

**Issue by kaishxu** [jingtaozhan/RepBERT-Index#5]
Hello, I notice you try to mask the attention between the query and document. However, it seems to be a problem when the input sequence is padded.
```python
def _mask_both_directions(valid_mask, token_type_ids):
    assert valid_mask.dim() == 2
    attention_mask = valid_mask[:, None, :]

    type_attention_mask = torch.abs(token_type_ids[:, :, None] - token_type_ids[:, None, :])
    attention_mask = attention_mask - type_attention_mask
    attention_mask = torch.clamp(attention_mask, 0, None)
    return attention_mask
```
e.g., if "token_type_ids" equals [0, 0, 0, 1, 1] (without padding),
then the "type_attention_mask" will be
|0 0 0 1 1|
|0 0 0 1 1|
|0 0 0 1 1|
|1   1  1  0  0|
|1   1  1  0  0|
and "attention_mask" will be
|1 1 1 0 0|
|1 1 1 0 0|
|1 1 1 0 0|
|0 0 0 1 1|
|0 0 0 1 1|
it is correct.
However, if "token_type_ids" equals [0, 0, 0, 1, 1, 0] (with padding, max_length=6),
then the "type_attention_mask" will be
|0 0 0 1 1 0|
|0 0 0 1 1 0|
|0 0 0 1 1 0|
|1 1 1 0 0 1|
|1 1 1 0 0 1|
|0 0 0 1 1 0|
and "attention_mask" will be
|1 1 1 0 0 0|
|1 1 1 0 0 0|
|1 1 1 0 0 0|
|0 0 0 1 1 0|
|0 0 0 1 1 0|
|1 1 1 0 0 0|
it is not the expected attention mask, right?

**Comment by jingtaozhan** [comment]
Your are right except for the first sentence and the last sentence. It is the expected attention mask.
|1 1 1 0 0 0|
|1 1 1 0 0 0|
|1 1 1 0 0 0|
|0 0 0 1 1 0|
|0 0 0 1 1 0|
|1 1 1 0 0 0|
We don't care the last row, since it doesn't affect the representations of the meaningful tokens.
And the last column is all zero as we expect.
